source/common/router/config_impl.h

```diff
@@ -787,7 +787,13 @@ class RouteEntryImplBase : public RouteEntryAndRoute,
   std::string newUri(const Http::RequestHeaderMap& headers) const override;
   void rewritePathHeader(Http::RequestHeaderMap&, bool) const override {}
   Http::Code responseCode() const override { return direct_response_code_.value(); }
-  const std::string& responseBody() const override { return direct_response_body_; }
+  const std::string& responseBody() const override {
+    absl::MutexLock lock(&direct_response_mutex_);
+    if (!direct_response_body_) {
+      return EMPTY_STRING;
+    }
+    return *direct_response_body_.get();
+  }
```

Comment:
Lock is expensive, for the body from inline string or inline bytes, the lock is unnecessary.

I think the lock here is used to ensure the access to the direct_response_body_ is thread safe, you still return a reference to the underlying string, then the lock make no any sense.
Please return a copy to the shared pointer of the body string.

Reply:
```
  std::shared_ptr<const string> responseBody() const override {
    if(direct_response_file_watcher_ == nullptr) {
      return direct_response_body_;
    }
    absl::MutexLock lock(&direct_response_mutex_);
    return direct_response_body_;
  }
```

Reply:
+1, no locks on worker threads please.
The general paradigm for Envoy is the main thread should have file watchers, then send dispatcher.post() updates to worker threads which swap out worker thread pointers in the worker thread context.

Reply:
The problem here is there is no worker thread specific pointers. We moved away from TLS to use this shared pointer model. But I think @wbpcode main point is about not introducing locks for other types direct response body (inline, static file etc) which makes sense and will change. Am I misunderstanding some thing here?

Reply:
Ideally, I think actomic shared pointer is best choice (better than TLS). I want to replace it with atomic shared_ptr once c
++20 is ready. So, the lock is temporarily compromise.

But if our rules cannot accept this lock completely, then seems we can only use the TLS.

Reply:
@alyssawilk based on this https://github.com/envoyproxy/envoy/pull/32789#discussion_r1533285639 comment, we moved away from TLS to this as a temporary measure till we get atomic shared ptr. This would be for a special case of a direct response body that user wishes it to be dynamic (I will introduce that API change) once we align on this locking behaviour. Please suggest

Reply:
Make things work perfectly is not easy.

If the TLS is the community's requirement, (and you (@ramaraochavali ) don't want to pay more time on the details of TLS, and if you can accept it,) I can try to create a general `DataSourceProvider` these two days (in a separated PR or your current PR directly, up to you), the provider will wrap all the details of TLS.
Then you only need to use the provider directly to complete the remaining work.

Reply:
Per offline discussion, will try to create a general data source provider. :)